module/extension/Director/src/director/find_ok_solution.cpp

```diff
@@ -0,0 +1,103 @@
+                return o;
+            }
+        }
+
+        // We couldn't find an OK solution, return blocked
+        return OkSolution(true, std::move(blocking_groups));
+    }
+
+    Director::OkSolution Director::find_ok_solution(const Solution::Option& option,
+                                                    std::set<std::type_index> used_types) {
+
+        // If the passed option is not an OK option it is blocked
+        if (option.state != Solution::Option::OK) {
+            return OkSolution(true, {option.provider->type});
+        }
+
+        // If this type was previously used by another branch we can't reuse it
+        // This can happen when you try to run two sibling tasks that have mutual needs
+        if (option.provider != nullptr && !used_types.insert(option.provider->type).second) {
+            return OkSolution(true, {option.provider->type});
+        }
+
+        // Store the providers that are needed for each of our requirements
+        std::vector<std::shared_ptr<Provider>> requirement_providers;
+
+        // If we have requirements we need to combine their output as our solution
+        bool blocked = false;
+        std::set<std::type_index> blocking_groups;
+        for (const auto& r : option.requirements) {
+
+            // If this is a pushed requirement we need to watch it
+            if (r.pushed) {
+                blocked = true;
+                blocking_groups.insert(option.provider->type);
+            }
+            else {
+                // Choose the best option for this requirement
+                auto s = find_ok_solution(r, used_types);
+
+                blocked = blocked || s.blocked;
+
+                // Add the blocked types to our blocked set
+                blocking_groups.insert(s.blocking_groups.begin(), s.blocking_groups.end());
+
+                // Condense the providers in this requirement into a single list
+                requirement_providers.push_back({s.provider});
+                requirement_providers.insert(requirement_providers.end(), s.requirements.begin(), s.requirements.end());
+
+                // Add the used types to our list
+                used_types.insert(s.used.begin(), s.used.end());
+            }
+        }
+
+        return OkSolution(blocked,
+                          option.provider,
+                          std::move(requirement_providers),
+                          std::move(used_types),
+                          std::move(blocking_groups));
+    }
+
+    std::vector<Director::OkSolution> Director::find_ok_solutions(const std::vector<Solution>& solutions) {
+
+        // Find each one individually but pass through the used types
+        std::vector<OkSolution> ok_solutions;
+        std::set<std::type_index> used_types;
+        for (const auto& solution : solutions) {
+
+            // Choose an option
+            ok_solutions.push_back(find_ok_solution(solution, used_types));
```

Comment:
This is hurting my brain a bit. So we have local set of `used_types`, which accumulates the used_types from the ok_solutions, but then we pass that set into the find_ok_solution function for the next call? This seems very odd (but it might be deliberate).

Reply:
So as you go down the options, you use types that can't be used again. However, when you are traversing down multiple options each option isn't going to be interfering with each other so they each get a copy of the current state.
When you've selected an option it gets passed down to the sibling solutions to limit them.

Yes it does hurt the brain a little :P

Reply:
Yep, my brain exploded